Evaluating the existing recommenders

florian opened the pull request:
This pull request adds a Jupyter notebook that evaluates how often the existing recommenders can be used.

Dexterp37 commented:
@mlopatka, would you kindly review this notebook as well? Ignore the "ipynb" file and leave your comments on the markdown version of the notebook.

mlopatka commented on evaluation.md:
```diff
@@ -0,0 +1,413 @@
+    scalar_parent_browser_engagement_tab_open_event_count[1].value as tab_open_count,
+    places_bookmarks_count[1].sum as bookmark_count,
+    scalar_parent_browser_engagement_unique_domains_count[1].value as unique_tlds
+FROM longitudinal
```
May want to add a filter to ensure only valid Fx clients end up in here... there is a lot of wierdness in telemetry.

`WHERE normalized_channel='release'
          AND build IS NOT NULL
          AND build[0].application_name='Firefox'`

could do the trick.

mlopatka replied:
This does not appear to have been addressedin the latest commit. Please double check that only valid FX clients are being included.

florian replied:
Sorry for that. I added it in the beginning but forgot to re-add it when changing the query for Hive.

mlopatka commented on evaluation.md:
````diff
@@ -0,0 +1,413 @@
+
+```python
+def format_frequency(frequency):
+    return "%.3f%%" % (frequency  * 100)
````
don't multiply by 100 here, just show the raw numbers as they occured in your sample.

mlopatka commented on evaluation.md:
````diff
@@ -0,0 +1,413 @@
+```python
+class DummySimilarityRecommender:
+    def can_recommend(self, client_data):
+        REQUIRED_FIELDS = ["geoCity", "subsessionLength", "locale", "os", "bookmark_count", "tab_open_count",
````
nit: some whitespace issues here in the nptebook version.

florian replied:
I guess this was because I copied that part from the taar repository. Should be fixed now.

mlopatka commented on evaluation.md:
```diff
@@ -0,0 +1,413 @@
+def complete_client_data(client_data):
+    client = client_data.asDict()
+    
+    client['installed_addons'] = client['installed_addons'] or []
```
Here can we filter out legacy addons and system addons? I speculate that the high rate of "can_recommend" positives you see for collaborative recommender is because it expects a more rigorously filtered data source. CollaborativeRecommender().can_recommend is only looking at `len(client_data.get('installed_addons', [])) > 0:`

Can you double check that only valid addons that can be used by collaborative recommender are included in the client sample you pass to it in this notebook.

florian replied:
I'm filtering out system addons now. This reduced the number for the collaborative filter to ~40%.

mlopatka commented on evaluation.md:
```diff
@@ -0,0 +1,413 @@
+          data=data
+)
+
+sorted_dataframe(df, results.values())
```
Please separate the rows where **only** one recommender is capable of making a recommendation. These results would be a lot easier to interpret is the tables were grouped by the number of `True` values in the row.

florian commented:
System addons are now filtered out.

I changed the formatting of the tables a little bit. They are much more readable if I use empty cells when recommenders couldn't make a recommendation. The exact behaviour for this is configurable in the notebook.

I also added a new section in the bottom of the notebook where the results are grouped by the number of recommenders that could make recommendations.
